### PR TITLE
[runtime] Add a signal handler for sigterm

### PR DESCRIPTION
**Pull request by alexanderkyte**
Fixes https://github.com/mono/mono/issues/8236

### CODE REVIEW

**Comment by alexanderkyte** [comment]
Fixes https://github.com/mono/mono/issues/8236

**Comment by alexanderkyte** [comment]
@monojenkins build pkg

**Comment by kumpera** [comment]
What's the impact will that have on regular usage of ctrl+c from CLI usage?

**Comment by vargaz** [comment]
Yes, users expect ctlr+c to just terminate the app without doing anything else.

**Comment by alexanderkyte** [comment]
Ctrl+c sends a sigint, sigterm is sent by Force Quit and by POSIX `kill` .

**Comment by alexanderkyte** [comment]
SIGTERM is like SIGKILL, but the dying process is allowed resources to shut down cleanly. This was requested by VS4M for detecting force quit of applications due to hangs.

**Comment by alexanderkyte** [comment]
@monojenkins build pkg

File: mono/mini/mini-posix.c [hunk 1]
```diff
@@ -393,6 +402,13 @@ mono_runtime_posix_install_handlers (void)
 
 	add_signal_handler (SIGABRT, sigabrt_signal_handler, 0);
 
+	gchar *sigterm_env = g_getenv ("MONO_HANDLE_SIGTERM");
```
Comment (alexrp) [inline mono/mini/mini-posix.c]: This should be documented in `man/mono.1`.

Comment (luhenry) [reply mono/mini/mini-posix.c]: Use `g_hasenv`

**Comment by luhenry** [comment]
@lewurm if we want to keep this feature for MERP only, then we will just enable it in the same icall as we enable MERP support, no need to create another one.

File: mono/mini/mini-posix.c [hunk 2]
```diff
@@ -210,6 +211,20 @@ MONO_SIG_HANDLER_FUNC (static, sigabrt_signal_handler)
 	}
 }
 
+MONO_SIG_HANDLER_FUNC (static, sigterm_signal_handler)
+{
+	MONO_SIG_HANDLER_INFO_TYPE *info = MONO_SIG_HANDLER_GET_INFO ();
+	MONO_SIG_HANDLER_GET_CONTEXT;
+
+#ifdef HOST_DARWIN
```
Comment (lewurm) [inline mono/mini/mini-posix.c]: this should be `TARGET_OSX`

File: mono/mini/mini-posix.c [hunk 3]
```diff
@@ -210,6 +211,20 @@ MONO_SIG_HANDLER_FUNC (static, sigabrt_signal_handler)
...
+	if (mono_merp_enabled ())
+		mono_handle_native_crash ("SIGTERM", ctx, info);
+#endif
+
+	mono_chain_signal (MONO_SIG_HANDLER_PARAMS);
+	exit (1);
```
Comment (luhenry) [inline mono/mini/mini-posix.c]: Why keeping the `exit (1);`?

Comment (alexanderkyte) [reply mono/mini/mini-posix.c]: Without an exit(1), you won't exit? 
I think there's a minor API confusion here. mono_chain_signal only calls another signal handler if it's already been registered. There's no other sigterm handler, so if you put a print between the chain signal and the exit call, it fires. 

```Before chain signal``` And
``` After chain signal, before exit```

Both fire.

This corresponds to the expected state in add_signal_handler, when ```previous_sa.sa_flags & SA_SIGINFO``` is false, and we have no already-registered signal handler to save. 

For this to work the way you want it to, where removing `exit` doesn't make it not `exit`, you'd need to explicitly register the handler at the start of every run that makes mono `exit` when sent sigaction. POSIX just defaults to externally killing a process when there's no signal handler, that doesn't mean that there's automatically a saved signal handler.

**Comment by luhenry** [comment]
It's broken on Android and Windows

**Comment by luhenry** [comment]
@monojenkins build Windows x64

**Comment by luhenry** [comment]
@monojenkins build Linux x64

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64

**Comment by luhenry** [comment]
@monojenkins backport to 2018-02

**Comment by luhenry** [comment]
@monojenkins backport to 2018-04

**Pull request merged**